torch/csrc/distributed/c10d/init.cpp

```diff
@@ -2985,6 +2990,27 @@ such as `dist.all_reduce(tensor, async_op=True)`.
                   However, if timeout is set, it will block the CPU thread until the NCCL work is completed
                   or timed out. If timeout, exception will be thrown.
             )")
+      .def(
+          "get_future_result",
+          [](::c10d::Work& work) -> std::shared_ptr<jit::PythonFutureWrapper> {
+            return std::make_shared<jit::PythonFutureWrapper>(
+                work.getFutureResult());
+          },
+          R"(
+            Returns:
+                A ``torch.futures.Future`` object of int type which maps to the enum type of WorkResult
+                As an example, a future object can be retrieved
+                by ``fut = process_group.allreduce(tensor).get_future_result()``.
+
+            Example::
+                users can use ``fut.wait()`` to blocking wait for the completion of the work and
+                get the WorkResult by ``fut.value()``.
+                Also, users can use ``fut.then(call_back_func)`` to register a callback function to be called
+                when the work is completed, without blocking the current thread.
```

Comment:
which thread will the callback run on?

Reply:
Good question, usually the callback would be run on the thread which marks the future as 'complete', e.g., (https://www.internalfb.com/code/fbsource/[9d20cc497d9438a63edfc443cdb319936f008d66]/fbcode/caffe2/aten/src/ATen/core/ivalue_inl.h?lines=962). But if the future is already completed when users registers the callback, the callback function would be ran on the same thread where usr registration happens